sdk/provisioning/Azure.Provisioning.CloudMachine/src/Azd.cs

```diff
@@ -0,0 +1,141 @@
+    {
+        Directory.CreateDirectory(infraDirectory);
+
+        cmi.Build().Save(infraDirectory);
+        var cmid = ReadOrCreateCmid();
+
+        // main.bicep
+        var location = new BicepParameter("location", typeof(string));
+        var principalId = new BicepParameter("principalId", typeof(string));
+
+        ResourceGroup rg = new(nameof(rg), ResourceGroupVersion)
+        {
+            Name = cmid,
+            Location = location
+        };
+
+        Infrastructure mainBicep = new("main")
+        {
+            TargetScope = "subscription"
+        };
+        ModuleImport import = new("cm", $"cm.bicep")
+        {
+            Name = "cm",
+            Scope = new IdentifierExpression(rg.ResourceName)
+        };
+        import.Parameters.Add(nameof(location), location);
+        import.Parameters.Add(nameof(principalId), principalId);
+
+        mainBicep.Add(rg);
+        mainBicep.Add(import);
+        mainBicep.Add(location);
+        mainBicep.Add(principalId);
+        mainBicep.Build().Save(infraDirectory);
+
+        WriteMainParametersFile(infraDirectory);
+    }
+
+    private static void WriteMainParametersFile(string infraDirectory)
+    {
+        File.WriteAllText(Path.Combine(infraDirectory, $"{MainBicepName}.parameters.json"),
+        """
+        {
+          "$schema": "https://schema.management.azure.com/schemas/2019-04-01/deploymentParameters.json#",
+          "contentVersion": "1.0.0.0",
+          "parameters": {
+            "environmentName": {
+              "value": "${AZURE_ENV_NAME}"
+            },
+            "location" : {
+              "value" : "${AZURE_LOCATION}"
+            },
+            "principalId": {
+              "value": "${AZURE_PRINCIPAL_ID}"
+            }
+          }
+        }
+        """);
+    }
+
+    internal static string ReadOrCreateCmid()
+    {
+        string appsettings = Path.Combine(".", "appsettings.json");
+
+        string? cmid;
+        if (!File.Exists(appsettings))
+        {
+            cmid = GenerateCloudMachineId();
+
+            using FileStream file = File.OpenWrite(appsettings);
+            Utf8JsonWriter writer = new Utf8JsonWriter(file);
+            writer.WriteStartObject();
+            writer.WritePropertyName("CloudMachine"u8);
+            writer.WriteStartObject();
+            writer.WriteString("ID"u8, cmid);
+            writer.WriteEndObject();
+            writer.WriteEndObject();
+            writer.Flush();
+            return cmid;
+        }
+
+        using FileStream json = File.OpenRead(appsettings);
+        JsonElement je = JsonDocument.Parse(json).RootElement;
+        // attempt to read CM configuration from existing configuration file
+        if (je.TryGetProperty("CloudMachine"u8, out JsonElement cm))
+        {
+            if (!cm.TryGetProperty("ID"u8, out JsonElement id))
+            {
+                throw new NotImplementedException();
```

Comment:
Is it worth adding a message to the exception?

Reply:
I will at some point, but I don't even want NYI exception type. I would like to have parity with the config provided behavior.